checks/apps/quantumespresso/quantumespresso_check.py

```diff
@@ -1,43 +1,86 @@
-    variant = parameter(['maint', 'prod'])
-    modules = ['QuantumESPRESSO']
-    executable = 'pw.x'
-    executable_opts = ['-in', 'ausurf.in', '-pd', '.true.']
-    extra_resources = {
-        'switches': {
-            'num_switches': 1
+sys.path.append(str(pathlib.Path(__file__).parent.parent.parent / 'mixins'))
+from extra_launcher_options import ExtraLauncherOptionsMixin
+from sarus_extra_launcher_options import SarusExtraLauncherOptionsMixin
+from cuda_visible_devices_all import CudaVisibleDevicesAllMixin
+
+qe_tests = {
+    # R:total number of MPI ranks, T:threads per rank, P:walltime
+    'Au-surf': {
+        'zen2': {
+            # zen2 cpu nodes: 2sockets, 8 numa, 16c/numa = 128c (no MT)
+            'energy_reference': -11427.09017218,
+            'performance_reference': [{'R': 32, 'T': 4, 'P': 102.0}]
+        },
+        'zen3': {
+            # A100 nodes: 1socket, 4 numa, 16c/numa = 64c (no MT)
+            'energy_reference': -11427.09017218,
+            'performance_reference': [{'R': 4, 'T': 16, 'P': 32.0}]
         }
     }
-    strict_check = False
-    maintainers = ['LM']
-    tags = {'scs'}
+}
+
+class QuantumESPRESSOBase(rfm.RunOnlyRegressionTest):
+    energy_tolerance = 1.0e-6
+    executable = 'pw.x'
+    # TODO: tests should all have pw.in as input file
+    executable_opts = ['-in', 'ausurf.in']
 
     @run_after('init')
-    def set_prog_envs_and_tags(self):
-        if self.current_system.name in ['eiger', 'pilatus']:
-            self.valid_prog_environs = ['cpeIntel']
-        else:
-            self.valid_prog_environs = ['builtin']
+    def set_description(self):
+        self.descr = f'QuantumESPRESSO {self.test_name} check'
 
-        self.tags |= {
-            'maintenance' if self.variant == 'maint' else 'production'
-        }
+    @run_before('run')
+    def set_parallel_resources(self):
+        self.skip_if_no_procinfo()
+        processor_info = self.current_partition.processor
+        # zen2, zen3, etc.
+        self.cpu_label = processor_info.arch
+        ## device label, for example 4x-gpu-sm_80
+        #dev_label = ''
+        #if self.current_partition.devices:
+        #    for e in self.current_partition.devices:
+        #        dev_label = dev_label + str(e.num_devices) + 'x-' + e.type + '-' + e.arch
+        #self.dev_label = dev_label
+        # for now use cpu label to identify node type
+        self.node_label = self.cpu_label
+
+        # number of physical cores
+        num_cores = int(processor_info.num_cpus / processor_info.num_cpus_per_core)
+
+        self.ref_dict = (
+            qe_tests[self.test_name][self.node_label]
+        )
+        # total number of ranks
+        self.num_tasks = self.ref_dict['performance_reference'][0]['R']
+        # threads / rank
+        T = self.ref_dict['performance_reference'][0]['T']
+        self.num_cpus_per_task = T
+        # ranks per node
+        self.num_tasks_per_node = int(num_cores / T)
+        if not self.env_vars:
```

Comment:
not sure what is the `if` for here ?

Reply:
Is `self.env_vars` always available? Is yes, then if is not needed indeed.